### ISSUE TITLE
Clarify if adjacent colors within a component can be used for UI control  and focus identification

### ISSUE THREAD

**Issue by mraccess77** [w3c/wcag#1717]
For UI controls, the understanding document indicates that SC doesn't require color changes within the component but with the background colors adjacent to the component.  But what if the author wants to rely on that as the way to meet the SC- can they?  if there was a situation where the focus indicator did not meet adjacent contrast with the background but did have sufficient contrast with the component could it still meet SC 1.4.11 - I think it should as in my code pen example there is 21:1 contrast with the outline and the component but only 1.5:1 with the page background.

https://codepen.io/mraccess/pen/ExZvEXB

**Comment by alastc** [comment]
Hi Jon, are you talking about 1.4.11, or the new focus-appearance?

**Comment by mraccess77** [comment]
@alastc SC 1.4.11 - since it came up today related to the new SC and some folks said 1.4.11 adjacent contrast was only to the background and not the component - but previously you had said that it didn't specify.   Knowing what's not or is covered in SC 1.4.11 may help us make the case for the proposed SC 2.4.11 being needed.

**Comment by alastc** [comment]
What we have in the [understanding doc](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html#figure-two-star-ratings-failing) (which I was struggling to remember today) is:

> In combination with 2.4.7 Focus Visible, the visual focus indicator for a component must have sufficient contrast against the adjacent background when the component is focused, except where the appearance of the component is determined by the user agent and not modified by the author.
> 
> Note that this Success Criterion does not directly compare the focused and unfocused states of a control - if the focus state relies on a change of color (e.g., changing only the background color of a button), this Success Criterion does not define any requirement for the difference in contrast between the two states. 

So for your example, I would say: no. 

I don't remember saying "it didn't specify", to the effect that you can choose. Perhaps I meant that it could have been taken as either to start with (given the variety of scenarios), and we then worked out the best way to apply it?

To make it more extreme, you could change the background to `#eee`, so the contrast under 2:1 and it's barely visible.
If you could choose what the focus indicator contrasts with it is unhelpful, if it is the background to the element then at least it is sometimes helpful. 

I think David mentioned as well, that as the focus of the "User Interface Components" section is the UIC, it would need to be colours adjacent to that UIC, not within it.

**Comment by mraccess77** [comment]
It just seems odd that we have a focus indicator with 21:1 contrast yet it fails the SC 1.4.11 but would pass the more strict proposed SC 2.4.11.  This then seems to say that to meet both the current and proposed SC you'd have to have sufficient contrast both with the background and internally as well unless you have a 2px indicator which I had and but you can still fail the current SC 1.4.11.

**Comment by JAWS-test** [comment]
> It just seems odd that we have a focus indicator with 21:1 contrast yet it fails the SC 1.4.11 but would pass the more strict proposed SC 2.4.11

I do not see that 2.4.11 is fulfilled, because the contrast between blue and white is not sufficient with 1.5:1

**Comment by mraccess77** [comment]
@JAWS-test the current proposal of SC 2.4.11 in the editor's draft indicates that the adjacent contrasting colors are within the component - you would not compare the white to blue.

"The contrasting area also has a contrast ratio of least 3:1 against adjacent colors in the focused component,"

**Comment by JAWS-test** [comment]
@mraccess77 But we have also

> Contrasting area: There is an area of the focus indicator that has a contrast ratio of at least 3:1 between the colors in the focused and unfocused states.

In my opinion, this is not fulfilled

**Comment by mraccess77** [comment]
@JAWS-test I believe "Contrasting area: There is an area of the focus indicator that has a contrast ratio of at least 3:1 between the colors in the focused and unfocused states."  

I think this only applies within the component though and this outline is outside the component - but I guess that is a questions of where does the component begin and end.

**Comment by JAWS-test** [comment]
@mraccess77 

> I think this only applies within the component though and this outline is outside the component 

I don't think so. First of all, it is not mentioned in the SC. Secondly, it would not make sense, because the focus indicator can be inside or outside the interactive element - but it must always be clearly visible.

**Comment by mraccess77** [comment]
I think we need more guidance on SC 2.4.11 in terms of comparing the focused and unfocused state if that state lies outside of the bounds of the component such as with an outset border.  What I'm afraid of is that we have to have contrast on both the outside edge, inside edge, and also with the previous colors and we could end up in a situation that is very difficult to pass.

**Comment by JAWS-test** [comment]
>  What I'm afraid of is that we have to have contrast on both the outside edge, inside edge, and also with the previous colors 

In my opinion, this is exactly what is required by 2.4.11. That is why there are the two contrast requirements:

> Contrasting area: There is an area of the focus indicator that has a contrast ratio of at least 3:1 between the colors in the focused and unfocused states.
> Adjacent contrast: The contrasting area also has a contrast ratio of least 3:1 against adjacent colors in the focused component, or the contrasting area has a thickness of at least 2 CSS pixels.

**Comment by alastc** [comment]
> It just seems odd that we have a focus indicator with 21:1 contrast yet it fails the SC 1.4.11 but would pass the more strict proposed SC 2.4.11. 

I'm not sure what you're scenario is there, it didn't line up with the codepen example?

> This then seems to say that to meet both the current and proposed SC you'd have to have sufficient contrast both with the background and internally as well unless you have a 2px indicator which I had and but you can still fail the current SC 1.4.11.

It depends on where your indicator is, and whether the component itself has contrast with the background. Something inside the component doesn't trigger 1.4.11, and something outside (with spacing from the component) doesn't trigger the adjacent contrast of 2.4.11.

We are covering a lot of scenarios, which is why the newer SC has more requirements.

> "Contrasting area: There is an area of the focus indicator that has a contrast ratio of at least 3:1 between the colors in the focused and unfocused states."... I think this only applies within the component though and this outline is outside the component - but I guess that is a questions of where does the component begin and end.

Adjacency isn't relevant here, it is the difference between focused and unfocused. That might be in the component, next to it, out of it, technically it could be somewhere else entirely (although that is not encouraged!)

Remember this used to be called "change of contrast".

We can work on the understanding doc, but essentially:

- Change of contrast does not check adjacency at all.
- 1.4.11 applies if the focus indicator is adjacent to the background of the component.
- 2.4.11 adjacent contrast applies to the indicator and component. If something is internal to the component, then that is essentially the same requirement (it doesn't clash).

That's needed because components themselves may or may not contrast with the background. E.g. a text link probably doesn't have a boundary that contrasts, so the contrast requirement across SCs boils down to that outline having contrast with the background.

However, if your component _does_ contrast with the background, that's where adjacent contrast is really needed, otherwise you can get the 1px black outline around a black box.

**Comment by mraccess77** [comment]
@alastc  "I'm not sure what you're scenario is there, it didn't line up with the codepen example?"

The black button has a white focus indicator that has 21:1 contrast between the white and black.

**Comment by alastc** [comment]
Ok, to answer this bit then:
> It just seems odd that we have a focus indicator with 21:1 contrast yet it fails the SC 1.4.11 but would pass the more strict proposed SC 2.4.11.

A couple of people commented that they wanted to minimise overlap, so if something fails 1.4.11 it shouldn't (necessarily) fail 2.4.11. That's why the adjacent contrast bullet was changed to make it with the component, it used to be all adjacent colors.

**Comment by mraccess77** [comment]
> A couple of people commented that they wanted to minimise overlap, so if something fails 1.4.11 it shouldn't (necessarily) fail 2.4.11. That's why the adjacent contrast bullet was changed to make it with the component, it used to be all adjacent colors.

Any examples we then look at must take both criteria into account since both will have to be met and both test different things.  When combined I think it will require a double focus indicator to pass - but now I am concerned that a double focus indicator with white and black could also actually fail.

**Comment by mraccess77** [comment]
> It depends on where your indicator is, and whether the component itself has contrast with the background. Something inside the component doesn't trigger 1.4.11, and something outside (with spacing from the component) doesn't trigger the adjacent contrast of 2.4.11.

Seems that the most common scenario is with a border without space that touches both the page background and the component or a change of background for the whole component which also touches the page background.

**Comment by mraccess77** [comment]
The case I find failing SC 1.4.11 but passing SC 2.4.11 is a 2px wide outset outline indicator where it doesn't contrast enough with adjacent colors on the page background.  For SC 2.4.11 the 2px allows it to pass - but SC 1.4.11 still fails against adjacent colors since the styles of the component were modified 
![image](https://user-images.githubusercontent.com/5640755/145082502-7abed9d9-3198-480c-9189-f052dbde5456.png).
